src/storage-client/src/controller/collection_mgmt.rs

```diff
@@ -288,46 +129,201 @@ impl CollectionManager {
+                                };
+
+                                match append_result {
+                                    // Everything was successful!
+                                    Ok(()) => {
+                                        // Notify all of our listeners.
+                                        notify_listeners(responders, || Ok(()));
+                                        // Break out of the retry loop so we can wait for more data.
+                                        break 'append_retry;
+                                    },
+                                    // Failed to write to some collections,
+                                    Err(StorageError::InvalidUppers(failed_ids)) => {
+                                        // It's fine to retry invalid-uppers errors here, since
+                                        // monotonic appends do not specify a particular upper or
+                                        // timestamp.
+
+                                        assert_eq!(failed_ids.len(), 1, "received errors for more than one collection");
+                                        assert_eq!(failed_ids[0], id, "received errors for a different collection");
+
+                                        // We've exhausted all of our retries, notify listeners
+                                        // and break out of the retry loop so we can wait for more
+                                        // data.
+                                        if retries.next().await.is_none() {
+                                            notify_listeners(responders, || Err(StorageError::InvalidUppers(vec![id])));
+                                            error!("exhausted retries when appending to managed collection {failed_ids:?}");
+                                            break 'append_retry;
+                                        }
+
+                                        debug!("Retrying invalid-uppers error while appending to managed collection {failed_ids:?}");
+                                    }
+                                    // Uh-oh, something else went wrong!
+                                    Err(other) => {
+                                        panic!("Unhandled error while appending to managed collection {id:?}: {other:?}")
+                                    }
+                                }
+                            }
+
+                            // Wait until our artificial latency has completed.
+                            //
+                            // Note: if writing to persist took longer than `DEFAULT_TICK` this
+                            // await will resolve immediately.
+                            tokio::time::sleep_until(min_time_to_complete).await;
+                        }
+                    }
+
+                    // If we haven't received any updates, then we'll move the upper forward.
+                    _ = interval.tick() => {
```

Comment:
I'm not sure if this would work, but would it be possible to always commit the data on `interval.tick()`? That way we never miss it and we don't try to commit too many things to persist per second

Reply:
Yeah I thought about that, it would make this logic a bit simpler, but the downside is it increases the latency of writes. For writes whose frequency is greater than tick interval, we should immediately send them to persist. If we only ever committed data on tick, then low frequency writes could end up with high latency when they don't need to

Reply:
Ah I see

Hopefully not too derailing: why don't we use the same logic we have for sources here? I'm obviously biased looking into the strict serializable stuff but the fewer places we mint timestamps differently the easier it is to analyze and convince ourselves that we can give strict serializable consistency without lag

Reply:
Not derailing at all!

Why don't we use the same logic as other sources is a great question! I don't know if there was a distinct reasoning (@sploiselle would know) but here appending data as soon as possible is relatively easy (compared to other sources) and at least naively it gives us better performance than waiting for an interval, so I imagine the reason might be "why not"? But if we find there are performance implications because of minting timestamps and strict seralizability, I'm sure we'd be more than happy to change the impl!